routing/result_interpretation.go

```diff
@@ -31,211 +31,211 @@ func interpretResult(rt *route.Route, failureSrcIdx *int,
+
+		reportOutgoing()
+
+	// If the next hop in the route wasn't known or offline, we'll only
+	// penalize the channel set which we attempted to route over. This is
+	// conservative, and it can handle faulty channels between nodes
+	// properly. Additionally, this guards against routing nodes returning
+	// errors in order to attempt to black list another node.
 	case *lnwire.FailUnknownNextPeer:
-		i.pairResults[failedPair] = 0
-		return false, 0
-
-	// If the node wasn't able to forward for which ever
-	// reason, then we'll note this and continue with the
-	// routes.
-	case *lnwire.FailTemporaryNodeFailure:
-		i.nodeFailures[failureVertex] = struct{}{}
-		return false, 0
-
-	case *lnwire.FailPermanentNodeFailure:
-		i.nodeFailures[failureVertex] = struct{}{}
-		return false, 0
-
-	// If we crafted a route that contains a too long time
-	// lock for an intermediate node, we'll prune the node.
-	// As there currently is no way of knowing that node's
-	// maximum acceptable cltv, we cannot take this
-	// constraint into account during routing.
-	//
-	// TODO(joostjager): Record the rejected cltv and use
-	// that as a hint during future path finding through
-	// that node.
-	case *lnwire.FailExpiryTooFar:
-		i.nodeFailures[failureVertex] = struct{}{}
-		return false, 0
-
-	// If we get a permanent channel or node failure, then
-	// we'll prune the channel in both directions and
-	// continue with the rest of the routes.
+		reportOutgoing()
+
+	// If we get a permanent channel, we'll prune the channel set in both
+	// directions and continue with the rest of the routes.
 	case *lnwire.FailPermanentChannelFailure:
-		i.pairResults[failedPair] = 0
-		i.pairResults[failedPair.Reverse()] = 0
-		return false, 0
+		reportOutgoing()
+
+	// If we get a failure due to violating the channel policy, we request a
+	// second chance because our graph may be out of date. An attached
+	// channel update should have been applied by now. If the second chance
+	// is granted, we try again. Otherwise either the error source or its
+	// predecessor sending an incorrect htlc is to blame.
+	case *lnwire.FailAmountBelowMinimum,
+		*lnwire.FailFeeInsufficient,
+		*lnwire.FailIncorrectCltvExpiry,
+		*lnwire.FailChannelDisabled:
+
+		// Set the node pair that is responsible for this failure. The
+		// second chance logic uses the policyFailure field.
+		i.policyFailure = &DirectedNodePair{
+			From: route.Hops[errorSourceIdx-1].PubKeyBytes,
+			To:   route.Hops[errorSourceIdx].PubKeyBytes,
+		}
+
+		// Assuming no second chance, we report incoming channel.
+		reportIncoming()
+
+	// If the outgoing channel doesn't have enough capacity, we penalize.
+	// But we penalize only in a single direction and only for amounts
+	// greater than the attempted amount.
+	case *lnwire.FailTemporaryChannelFailure:
+		reportOutgoingBalance()
 
-	// Any other failure or an empty failure will get the node pruned.
+	// If FailExpiryTooSoon is received, there must have been some delay
```

Comment:
Alternatively, this could actually be us being behind the chain (not yet synced).

Reply:
Interesting. The question is how to deal with this? Isn't the solution to not try to send payments if we are not yet synced? Because we can't calculate the proper htlc expiries?

Reply:
Yeah we do this atm if we're still syncing, but if we fall behind for w/e reason (RPC connection dies or something), then it isn't covered yet. Not saying we need to handle this here, just pointing it out for future consideration.

Reply:
Comment extended to describe this case and mention that we ignore it for now.